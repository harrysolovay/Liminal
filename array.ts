import { make, type Ty } from "./_base.ts"

export function array<E extends Ty>(description: string | undefined, element: E): ArrayTy<E> {
  return make((ctx) => ({
    type: "array",
    description,
    items: ctx.ref(element),
  }))
}

export type ArrayTy<E extends Ty = any> = Ty<Array<InstanceType<E>>>
