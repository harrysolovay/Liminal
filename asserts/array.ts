import { assert } from "../util/mod.ts"

export function minLength<T>(minLength: number, value: Array<T>) {
  assert(value.length >= minLength, `Length must be gte ${minLength} elements.`)
}

export function maxLength<T>(maxLength: number, value: Array<T>) {
  assert(value.length <= maxLength, `Length must be lte ${maxLength} elements.`)
}
