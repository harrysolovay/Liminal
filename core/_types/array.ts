import type { Type } from "../Type.ts"
import { declareType } from "./_declareType.ts"

export function array<T>(element: Type<T>): Type<Array<T>> {
  return declareType(() => array<T>, [element])
}
