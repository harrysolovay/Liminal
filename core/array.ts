import type { Type } from "../Type.ts"
import { declare } from "../TypeDeclaration.ts"
import { assert } from "../util/assert.ts"

export function array<E extends Type>(element: E): Type<Array<E["T"]>, {
  minLength: number
  maxLength: number
}, E["P"]> {
  return declare<Array<E["T"]>>()({
    name: "array",
    source: {
      factory: array,
      args: { element },
    },
    subschema: (ref) => ({
      type: "array",
      items: ref(element),
    }),
    visitor: (value, visit) => value.map((v, i) => visit(v, element, i)),
    assertRefinementsValid: ({ minLength, maxLength }) => {
      assert(
        !(typeof minLength === "number" && typeof maxLength === "number") || minLength <= maxLength,
      )
    },
    assertRefinements: {
      minLength: (value, minLength) => value.length >= minLength,
      maxLength: (value, maxLength) => value.length <= maxLength,
    },
  })
}
