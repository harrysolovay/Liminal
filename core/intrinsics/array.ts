import { declareType } from "../declareType.ts"
import type { Type } from "../Type.ts"

export function array<T, E>(element: Type<T, E>): Type<Array<T>, E> {
  return declareType({
    type: "array",
    self() {
      return array
    },
    args: [element],
  })
}
