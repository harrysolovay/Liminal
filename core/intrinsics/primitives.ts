import { assert } from "@std/assert"
import type { Type } from "../Type.ts"
import { declare } from "./declare.ts"

export { null_ as null }
const null_: Type<null> = declare({
  getAtom: () => null_,
  assert: (value) => {
    assert(value === null)
  },
})

export const boolean: Type<boolean> = declare({
  getAtom: () => boolean,
  assert: (value) => {
    assert(typeof value === "boolean")
  },
})

export const integer: Type<number> = declare({
  getAtom: () => integer,
  assert: (value) => {
    assert(Number.isInteger(value))
  },
})

export const number: Type<number> = declare({
  getAtom: () => number,
  assert: (value) => {
    assert(typeof value === "number")
  },
})

export const string: Type<string> = declare({
  getAtom: () => string,
  assert: (value) => {
    assert(typeof value === "string")
  },
})
