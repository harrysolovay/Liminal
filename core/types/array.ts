import { type AnyType, Type } from "../Type.ts"

export function array<E extends AnyType>(element: E): Type<Array<E["T"]>, E["P"]> {
  return Type({
    kind: "array",
    factory: array,
    args: [element],
    argsLookup: { element },
  })
}
