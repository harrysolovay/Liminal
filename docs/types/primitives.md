# Primitive Types

```ts twoslash
import { L } from "liminal"
// ---cut---
L.boolean
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.number
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.integer
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.string
// ^?
```

<br />
