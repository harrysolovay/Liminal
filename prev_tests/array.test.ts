import { assertTySnapshot } from "test_util"
import { array } from "./array.ts"
import { string } from "./primitives.ts"

Deno.test("array", async (t) => {
  await assertTySnapshot(t, array(string))
})
