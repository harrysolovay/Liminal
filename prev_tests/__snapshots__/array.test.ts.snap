export const snapshot = {};

snapshot[`array 1`] = `
{
  items: {
    type: "string",
  },
  type: "array",
}
`;

snapshot[`array 2`] = `
{
  description: "One.",
  items: {
    type: "string",
  },
  type: "array",
}
`;

snapshot[`array 3`] = `
{
  description: "Two. One.",
  items: {
    type: "string",
  },
  type: "array",
}
`;
